src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs

```diff
@@ -0,0 +1,155 @@
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        internal static readonly DiagnosticDescriptor CountDescriptor = DiagnosticDescriptorHelper.Create(
+            CountId,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyTitle)),
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            LengthDescriptor,
+            CountDescriptor,
+            IsEmptyDescriptor
+        );
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterOperationAction(OnInvocationAnalysis, OperationKind.Invocation);
+        }
+
+        private static void OnInvocationAnalysis(OperationAnalysisContext context)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var originalMethod = invocation.TargetMethod.OriginalDefinition;
+            if (originalMethod.MethodKind == MethodKind.ReducedExtension)
+            {
+                originalMethod = originalMethod.ReducedFrom;
+            }
+
+            if (IsEligibleAnyMethod(originalMethod))
+            {
+                var type = invocation.GetReceiverType(context.Compilation, beforeConversion: true, context.CancellationToken);
+                if (type is null)
+                {
+                    return;
+                }
+
+                if (HasEligibleIsEmptyProperty(type))
+                {
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(IsEmptyDescriptor));
+
+                    return;
+                }
+
+                if (HasEligibleLengthProperty(type))
+                {
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(LengthDescriptor));
+
+                    return;
+                }
+
+                if (HasEligibleCountProperty(type))
+                {
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(CountDescriptor));
+
+                    return;
+                }
+            }
+        }
+
+        private static bool IsEligibleAnyMethod(IMethodSymbol method)
+        {
+            return method is
+            {
+                Name: AnyText,
+                ReturnType.SpecialType: SpecialType.System_Boolean,
+                IsExtensionMethod: true,
+                Parameters: [_]
+            };
+        }
+
+        private static bool HasEligibleIsEmptyProperty(ITypeSymbol typeSymbol)
```

Comment:
I agree with the intent of this heuristic. If properties such as `IsEmpty`, `Length`, or `Count` exist on a collection type and they don't represent information about the number of elements in that collection then the type is out of step with most of .NET.

However, as I mentioned, we are _assuming_ that the type is a collection type. We don't actually check for that yet. I would suggest that we verify that the type our `Any` method applies to inherits from either `IEnumerable` or `IEnumerable<T>`.

Reply:
Consider a record and extension method pair like this:
```csharp
public record UniversityClass(
  Professor? Instructor,
  ImmutableArray<Student> TeachingAssistants,
  ImmutableArray<Student> Students)
{
  // meant to indicate that a class has no students signed up.
  public bool IsEmpty => Students.Length == 0;
}

// meant to indicate that a class has no instructors assigned to teach it.
public static bool Any(UniversityClass @class)
  => @class.Instructor is not null ||
     @class.TeachingAssistants.Length > 0;
```